test/handlebars.js

```diff
@@ -17,14 +17,15 @@ var tmpl = [
   'lower: {{lower name}}',
   'spacer: {{spacer name}}',
   'spacer-delim: {{spacer name "-"}}',
-  'lower(upper): {{lower (upper name)}}'
-  // 'spacer(upper, lower): {{spacer (upper name) (lower "X")}}',
-  // 'block: {{#block}}{{upper name}}{{/block}}',
-  // 'ifConditional1: {{#if (equals "foo" foo)}}{{upper name}}{{/if}}',
-  // 'ifConditional2: {{#if (equals "baz" bar)}}{{upper name}}{{/if}}',
-  // 'useHash: {{#useHash me=(lookup this "person")}}{{me.first}} {{me.last}}{{/useHash}}',
-  // 'sum: {{sum 1 2 3}}',
-  // 'lookup(this "person"): {{lookup this "person"}}'
+  'lower(upper): {{lower (upper name)}}',
+  'spacer(upper, lower): {{spacer (upper name) (lower "X")}}',
+  'block: {{#block}}{{upper name}}{{/block}}',
+  'ifConditional1: {{#if (equals "foo" foo)}}{{upper name}}{{/if}}',
```

Comment:
Pretty strange. I understand why it goes into the if. Because `equals` returns boolean false which is passed to the callback, but in reality, this means that it will get converted to string `'false'`, and that's why it goes inside.

Not sure how to handle this. You can see that the added `ifConditional3` is working correctly and does to go inside the if.

Reply:
I don't think we should handle such cases in the implementation, because it may break other cases such as lodash tests.

I think that in the `equals` helper we can check if it is falsey value then pass empty string or just call `cb()`

Reply:
The reason this happens is because of the 2 steps that are taken... the template is rendered into a string with async ids, then the ids are resolved. Since the `if` helper isn't wrapped, the async id is passed directly to the `if` logic so it looks like a truthy value.

This is the same reason why in engines like `lodash`, you can't directly use `if` statements with async helpers.

It's something that we've talked about and think that it would take another pass at the "engine" level to possible detect built-in helpers (like `{{#if}}` in `handlebars` or control flow statements like `if` in `lodash`. I don't think it should be handled in `async-helpers` since this is intended to be template engine agnostic.

Reply:
Aaaah. Yea, `equals` is async function so the return is again an id. 

> I don't think it should be handled in async-helpers since this is intended to be template engine agnostic.

Agree.

But how then it is working in the current master (v0.3)? If it's not a problem, we can continue.